anomaly-benchmarking/script.whizzml

```diff
@@ -0,0 +1,277 @@
+                       rest)))))))
+
+;;Two-Coloring
+
+;; Given a node id and a spanning tree, returns that node's neighbors
+;; as a list of the node ids
+
+(define (find-neighbors node span-tree)
+  (let (end-list (map (lambda (x) (get x "ends")) span-tree)
+        ends-with-node (filter (lambda (x) (member? node x)) end-list))
+    (filter (lambda (x) (not (= x node))) (flatten [] ends-with-node)))) 
+
+;; Given a list of nodes, a node id, and a color ("normal" or
+;; "anomalous"), returns the list with the color of the node changed
+
+(define (color-a-node nodes node color)
+  (let (old-node (head (filter (lambda (x) (= node (get x "id"))) nodes))
+        all-other-nodes (filter (lambda (x) (not (= old-node x))) nodes)
+        new-node (assoc old-node "color" color))
+    (append all-other-nodes new-node)))
+
+;; Given a graph and a spanning tree, returns a graph where the nodes
+;; have a new attribute "color", either normal or anomalous. No two
+;; touching nodes have the same color.
+
+(define (two-color graph span-tree)
+  (let (old-nodes (map (lambda (x) (assoc x "color" "none")) 
+                    (get graph "nodes"))
+        edges (get graph "edges")
+        uncolored? (lambda (id color-map) (= (color-map (str id)) "none"))
+        color-lots (lambda (nodes list-of-ids color) 
+                     (reduce (lambda (x y) (color-a-node x y color))
+                             nodes list-of-ids))
+        new-color (lambda (color) (if (= color "normal") 
+                                      "anomalous" "normal")))
+    (loop (queue [0] nodes (color-a-node old-nodes 0 "normal"))
+      (if (empty? queue)
+          (make-map ["nodes" "edges"] [nodes edges])
+          (let (node (head queue)
+                rest (tail queue)
+                node-color (mapify nodes "id" "color")
+                neighbors (filter (lambda (x) (uncolored? x node-color)) 
+                                  (find-neighbors node span-tree))
+                colored-nodes (color-lots nodes neighbors 
+                                         (new-color (node-color (str node))))
+                new-queue (concat rest neighbors))
+            (recur new-queue colored-nodes))))))
+
+;; Creating the new dataset
+
+;; Given a field id and a map of class names to colors, returns a
+;; flatline expression that returns a color based on the class in the
+;; given field.
+
+(define (color-filter field-id color-map)
+  (let (classes (keys color-map)
+        one-line (lambda (class color) 
+                   (flatline "(matches? (field {{field-id}}) {{class}}) {{color}}"))
+        conditions (reduce (lambda (x y) 
+                       (join " " [x (one-line y (color-map y))])) 
+                     "" (keys color-map)))
+    (join " " ["(cond" conditions ")"])))
+
+;; Given a dataset with a multivariate objective field, returns an
+;; extended dataset with a new field "binary", which can be either
```

Comment:
Super-nit:  Can we name the field something with a bit more semantics than "binary"?  Maybe "Is Anomoly?" or something?  I won't insist.